### PR TITLE
Fix race condition in CA1001 analyzer

### LINKED ISSUE #6780: Intermittent test failures for CA1001 in CI
> @CollinAlpert I am seeing intermittent test failures for CA1001 in CI after https://github.com/dotnet/roslyn-analyzers/pull/6741

Test: 
`Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.UnitTests.TypesThatOwnDisposableFieldsShouldBeDisposableAnalyzerTests.CA1001_ExcludedSymbolNamesAsync(editorConfigText: "")`

```
Expected diagnostic message to be "Type 'SomeClass' owns disposable field(s) '_fs', '_ms' but is not disposable" was "Type 'SomeClass' owns disposable field(s) '_fs' but is not disposable"


Expected diagnostic:
// /0/Test0.cs(4,14,4,23): hidden CA1001: Type 'SomeClass' owns disposable field(s) '_fs', '_ms' but is not disposable
VerifyCS.Diagnostic().WithSpan(4, 14, 4, 23).WithArguments("SomeClass", "_fs', '_ms"),


Actual diagnostic:
// /0/Test0.cs(4,14): hidden CA1001: Type 'SomeClass' owns disposable field(s) '_fs' but is not disposable
VerifyCS.Diagnostic().WithSpan(4, 14, 4, 23).WithArguments("SomeClass", "_fs"),


Assert.Equal() Failure
Expected: Type 'SomeClass' owns disposable field(s) '_fs', '_ms' but is not disposable
Actual:   Type 'SomeClass' owns disposable field(s) '_fs' but is not disposable



Stack trace
   at Microsoft.CodeAnalysis.Testing.Verifiers.XUnitVerifier.Equal[T](T expected, T actual, String message) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Testing.Verifiers.XUnit/XUnitVerifier.cs:line 49
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.VerifyDiagnosticResults(IEnumerable`1 actualResults, ImmutableArray`1 analyzers, DiagnosticResult[] expectedResults, IVerifier verifier) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 603
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.VerifyDiagnosticsAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, DiagnosticResult[] expected, IVerifier verifier, CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 465
   at Microsoft.CodeAnalysis.Testing.CodeFixTest`1.RunImplAsync(CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing/CodeFixTest`1.cs:line 305
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.RunAsync(CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 188
   at Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.UnitTests.TypesThatOwnDisposableFieldsShouldBeDisposableAnalyzerTests.CA1001_ExcludedSymbolNamesAsync(String editorConfigText) in /_/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/TypesThatOwnDisposableFieldsShouldBeDisposableTests.cs:line 875
--- End of stack trace from previous location ---
```

_Originally posted by @mavasani in https://github.com/dotnet/roslyn-analyzers/issues/6776#issuecomment-1639431028_

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #6780

### CODE REVIEW

**Comment by mavasani** [comment]
@CollinAlpert

**Review by sharwell — CHANGES_REQUESTED** [review]
Recommend using `TemporarySet<string>` for this.

**Comment by mavasani** [comment]
@sharwell Addressed feedback

File: src/Utilities/Compiler/PooledObjects/TemporarySet`1.cs [hunk 1]
```diff
@@ -74,11 +74,27 @@ public bool Add(T item, CancellationToken cancellationToken)
...
+
+            return storage.Count > 0;
+        }
+
         public readonly Enumerator GetEnumerator_NonConcurrent()
         {
             return new Enumerator((_storage ?? EmptyHashSet).GetEnumerator());
         }
 
+        public readonly IEnumerable<T> AsEunumerable_NonConcurrent()
```
Comment (sharwell) [inline src/Utilities/Compiler/PooledObjects/TemporarySet`1.cs:90]:
```suggestion
        public readonly IEnumerable<T> AsEnumerable_NonConcurrent()
```

File: src/Utilities/Compiler/PooledObjects/TemporarySet`1.cs [hunk 2]
```diff
@@ -74,11 +74,27 @@ public bool Add(T item, CancellationToken cancellationToken)
...
+        {
+            foreach (var item in _storage ?? EmptyHashSet)
```
Comment (sharwell) [inline src/Utilities/Compiler/PooledObjects/TemporarySet`1.cs:92]:
```suggestion
            foreach (var item in NonConcurrentEnumerable)
```

Comment (sharwell) [reply src/Utilities/Compiler/PooledObjects/TemporarySet`1.cs:92]: Actually I think I'd rather method just return directly:

```
return (_storage ?? EmptyHashSet).GetEnumerator();
```

Comment (mavasani) [reply src/Utilities/Compiler/PooledObjects/TemporarySet`1.cs:92]: Needs to be `return (_storage ?? EmptyHashSet).AsEnumerable();`. Done

**Review by sharwell — APPROVED** [review]

**Pull request merged**